Add type hints

jodal opened the pull request:
The goal here was only to add type hints, but on the way there I ended up refactoring quite a bit to make the types more helpful in guaranteeing that the right strings parsed out of the URLs are present, depending on the URL type.

All tests pass, including type checks with pyright and ty. I'm currently listening to Xmas songs in German from ORF :christmas_tree: 

@girst Please have a look, and if you find this too invasive, or the first commit too large, I can try to redo the refactoring in smaller steps.

jodal committed:
refactor: Use patterns that works well with type hints

This refactor is way too large, but I got carried away and it isn't very
tempting to split up this without redoing it.

jodal committed:
chore: Add annotations to all functions

jodal committed:
chore(ci): Enable pyright in CI

jodal committed:
chore: Add ty type checker

girst commented:
On Tue, Dec 23, 2025 at 08:48:00AM -0800, Stein Magnus Jodal wrote:
>All tests pass, including type checks with pyright and ty. I&#39;m 
>currently listening to Xmas songs in German from ORF :christmas_tree:

nice! the 'fm4' station is 50% english speaking, fyi!

>@girst Please have a look, and if you find this too invasive, or the 
>first commit too large, I can try to redo the refactoring in smaller 
>steps.

i'll not be able to do much until after Congress, sorry. but i will 
definitely make time for it afterwards. i really appreciate all the work 
you've been putting into mopidy in general, and -orfradio in particular!

girst commented:
I#ve finally made some time to review this patchset.
i'll have to be honest: i'm not exactly excited about *all* the changes 
(IMO a lot of typed code is more verbose and harder to read, especially 
when broken over multiple lines) - but I also don't really hate it. i do 
recognize that a well-typed codebase is a massive net positive! :)

>diff --git a/pyproject.toml b/pyproject.toml
>index 10909df..3c24352 100644
>--- a/pyproject.toml
>+++ b/pyproject.toml
>@@ -15,7 +15,7 @@ classifiers = [
> dynamic = ["version"]
> dependencies = [
>     "beaker",
>-    "mopidy >= 3.0",
>+    "mopidy >= 4.0.0a11",

do we really need the latest alpha version? i'm a bit behind on my 
fedora packages (alpha4, currently), and that makes reviewing this 
harder for me.

>     "pykka >= 4.1",
>     "python-dateutil",
> ]

other than that, i don't see anything that stands out as objectionable. 
i'd like to do some more live testing, and i'll merge it then.

again: big thank you!

jodal committed:
chore(deps): Drop mopidy required version to 4.0.0a4

Newer is better for types, but CI still pass on 4.0.0a4 and that eases
testing on Fedora.

jodal committed:
chore: Revert some typing changes

...to be compatible with the upcoming mopidy 4.0.0a14.

jodal commented:
I managed to get the CI passing with `==4.0.0a4` with just some very tiny changes, that anyway are in line with what will be in the next alpha release, 4.0.0a14. Thus, this PR should now be testable on 4.0.0a4 and be mergeable.

girst commented:
On Tue, Jan 13, 2026 at 12:02:23AM -0800, Stein Magnus Jodal wrote:
>jodal left a comment (mopidy/mopidy-orfradio#13)
>
>I managed to get the CI passing with `==4.0.0a4` with just some very 
>tiny changes, that anyway are in line with what will be in the next 
>alpha release, 4.0.0a14. Thus, this PR should now be testable on 
>4.0.0a4 and be mergeable.

thanks a lot! i'll likely get to it on the weekend!

girst reviewed (approved):
thanks, LGTM! the more i'm reading the changed code, the better i like it :)

do you want to rebase it once more or should i merge this as-is?

jodal commented:
Great! :-) I think we're already on top of `main`, so I'm just merging.
